File: services/mission-control/src/api/providerIntentSuggestions.ts
import { atlas } from "./atlas";
import type { ProviderMonitoringIntentSuggestionV1 } from "../types/providerManagement";

const fields = [
    "schema_version", "suggestion_id", "provider_id", "resource_type", "resource_id",
    "management_fingerprint", "suggested_expectation", "base_record_version", "source",
    "source_rule", "reason", "advisory_only", "grants_permission", "grants_execution",
] as const;
const fingerprintPattern = /^provider-management-fingerprint-v1:[a-f0-9]{64}$/;
const suggestionIdPattern = /^provider-monitoring-intent-suggestion-id-v1:[a-f0-9]{64}$/;
const resourceIdPattern = /^[1-9][0-9]*$/;

function parseSuggestion(value: unknown): ProviderMonitoringIntentSuggestionV1 {
    if (typeof value !== "object" || value === null || Array.isArray(value)) {
        throw new Error("Provider monitoring suggestion is invalid.");
    }
    const item = value as Record<string, unknown>;
    if (Object.keys(item).sort().join("|") !== [...fields].sort().join("|")) {
        throw new Error("Provider monitoring suggestion fields are invalid.");
    }
    if (item.schema_version !== "provider-monitoring-intent-suggestion-v1"
        || typeof item.suggestion_id !== "string" || !suggestionIdPattern.test(item.suggestion_id)
        || item.provider_id !== "proxmox" || item.resource_type !== "qemu"
        || typeof item.resource_id !== "string" || item.resource_id.length > 20 || !resourceIdPattern.test(item.resource_id)
        || typeof item.management_fingerprint !== "string" || !fingerprintPattern.test(item.management_fingerprint)
        || item.suggested_expectation !== "running" || item.base_record_version !== 0
        || item.source !== "provider_intelligence_rule"
        || item.source_rule !== "qemu-observed-running-no-active-intent-v1"
        || item.reason !== "observed_running_without_active_intent"
        || item.advisory_only !== true || item.grants_permission !== false
        || item.grants_execution !== false) {
        throw new Error("Provider monitoring suggestion contract is invalid.");
    }
    return {
        schema_version: item.schema_version,
        suggestion_id: item.suggestion_id as ProviderMonitoringIntentSuggestionV1["suggestion_id"],
        provider_id: item.provider_id,
        resource_type: item.resource_type,
        resource_id: item.resource_id,
        management_fingerprint: item.management_fingerprint as ProviderMonitoringIntentSuggestionV1["management_fingerprint"],
        suggested_expectation: item.suggested_expectation,
        base_record_version: item.base_record_version,
        source: item.source,
        source_rule: item.source_rule,
        reason: item.reason,
        advisory_only: item.advisory_only,
        grants_permission: item.grants_permission,
        grants_execution: item.grants_execution,
    };
}

export function parseProviderMonitoringIntentSuggestions(
    value: unknown,
): ProviderMonitoringIntentSuggestionV1[] {
    if (!Array.isArray(value)) throw new Error("Provider monitoring suggestions are invalid.");
    const parsed = value.map(parseSuggestion);
    const ids = new Set<string>();
    const coordinates = new Set<string>();
    for (const item of parsed) {
        const coordinate = `${item.provider_id}\u0000${item.resource_type}\u0000${item.resource_id}\u0000${item.source_rule}`;
        if (ids.has(item.suggestion_id) || coordinates.has(coordinate)) {
            throw new Error("Provider monitoring suggestions are duplicated.");
        }
        ids.add(item.suggestion_id);
        coordinates.add(coordinate);
    }
    const ordered = [...parsed].sort((left, right) => {
        const provider = left.provider_id.localeCompare(right.provider_id);
        if (provider !== 0) return provider;
        const type = left.resource_type.localeCompare(right.resource_type);
        if (type !== 0) return type;
        const leftResource = BigInt(left.resource_id);
        const rightResource = BigInt(right.resource_id);
        if (leftResource !== rightResource) return leftResource < rightResource ? -1 : 1;
        return left.suggestion_id.localeCompare(right.suggestion_id);
    });
    if (parsed.some((item, index) => item.suggestion_id !== ordered[index]?.suggestion_id)) {
        throw new Error("Provider monitoring suggestions are not canonically ordered.");
    }
    return parsed;
}

export async function getProviderMonitoringIntentSuggestions(
    providerId: string,
): Promise<ProviderMonitoringIntentSuggestionV1[]> {
    const response = await atlas.get<unknown>(
        `/providers/${encodeURIComponent(providerId)}/management/operator/monitoring-suggestions`,
        { withCredentials: true },
    );
    return parseProviderMonitoringIntentSuggestions(response.data);
}

File: services/atlas-core/app/models/provider_intent_suggestions.py
"""Strict advisory contracts for identity-bound Provider Intent suggestions."""

from __future__ import annotations

import hashlib
import json
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.provider_management import (
    PROVIDER_MANAGEMENT_FINGERPRINT_PATTERN,
    ProviderMonitoringExpectation,
)

SUGGESTION_SCHEMA_VERSION = "provider-monitoring-intent-suggestion-v1"
SUGGESTION_ID_VERSION = "provider-monitoring-intent-suggestion-id-v1"
OBSERVED_RUNNING_RULE = "qemu-observed-running-no-active-intent-v1"
OBSERVED_RUNNING_REASON = "observed_running_without_active_intent"
SUGGESTION_ID_PATTERN = (
    rf"^{SUGGESTION_ID_VERSION}:[a-f0-9]{{64}}$"
)


class ProviderIntentSuggestionModel(BaseModel):
    """Reject extension and mutation of advisory suggestion contracts."""

    model_config = ConfigDict(extra="forbid", frozen=True, strict=True)


def build_provider_monitoring_intent_suggestion_id(
    *,
    provider_id: str,
    resource_type: str,
    resource_id: str,
    management_fingerprint: str,
    suggested_expectation: ProviderMonitoringExpectation,
    base_record_version: int,
    source_rule: str,
    reason: str,
) -> str:
    """Bind one suggestion to its complete sanitized advisory source state."""

    payload = {
        "base_record_version": base_record_version,
        "management_fingerprint": management_fingerprint,
        "provider_id": provider_id,
        "reason": reason,
        "resource_id": resource_id,
        "resource_type": resource_type,
        "schema_version": SUGGESTION_SCHEMA_VERSION,
        "source_rule": source_rule,
        "suggested_expectation": suggested_expectation.value,
    }
    encoded = json.dumps(
        payload, sort_keys=True, separators=(",", ":")
    ).encode()
    return f"{SUGGESTION_ID_VERSION}:{hashlib.sha256(encoded).hexdigest()}"


class ProviderMonitoringIntentSuggestionV1(ProviderIntentSuggestionModel):
    """One immutable advisory produced by the accepted initial P5 rule."""

    schema_version: Literal[
        "provider-monitoring-intent-suggestion-v1"
    ] = SUGGESTION_SCHEMA_VERSION
    suggestion_id: str = Field(pattern=SUGGESTION_ID_PATTERN)
    provider_id: Literal["proxmox"]
    resource_type: Literal["qemu"]
    resource_id: str = Field(pattern=r"^[1-9][0-9]*$", max_length=20)
    management_fingerprint: str = Field(
        pattern=PROVIDER_MANAGEMENT_FINGERPRINT_PATTERN
    )
    suggested_expectation: Literal[
        ProviderMonitoringExpectation.RUNNING
    ]
    base_record_version: int = Field(ge=0)
    source: Literal["provider_intelligence_rule"]
    source_rule: Literal[
        "qemu-observed-running-no-active-intent-v1"
    ]
    reason: Literal["observed_running_without_active_intent"]
    advisory_only: Literal[True] = True
    grants_permission: Literal[False] = False
    grants_execution: Literal[False] = False

    @model_validator(mode="after")
    def validate_suggestion_identity(
        self,
    ) -> ProviderMonitoringIntentSuggestionV1:
        if self.base_record_version != 0:
            raise ValueError(
                "the initial suggestion rule requires base record version zero"
            )
        expected = build_provider_monitoring_intent_suggestion_id(
            provider_id=self.provider_id,
            resource_type=self.resource_type,
            resource_id=self.resource_id,
            management_fingerprint=self.management_fingerprint,
            suggested_expectation=self.suggested_expectation,
            base_record_version=self.base_record_version,
            source_rule=self.source_rule,
            reason=self.reason,
        )
        if self.suggestion_id != expected:
            raise ValueError(
                "suggestion ID does not match the advisory source state"
            )
        return self


def observed_running_suggestion(
    *, resource_id: str, management_fingerprint: str
) -> ProviderMonitoringIntentSuggestionV1:
    """Build the sole suggestion shape accepted by the initial P5 rule."""

    expectation = ProviderMonitoringExpectation.RUNNING
    suggestion_id = build_provider_monitoring_intent_suggestion_id(
        provider_id="proxmox",
        resource_type="qemu",
        resource_id=resource_id,
        management_fingerprint=management_fingerprint,
        suggested_expectation=expectation,
        base_record_version=0,
        source_rule=OBSERVED_RUNNING_RULE,
        reason=OBSERVED_RUNNING_REASON,
    )
    return ProviderMonitoringIntentSuggestionV1(
        suggestion_id=suggestion_id,
        provider_id="proxmox",
        resource_type="qemu",
        resource_id=resource_id,
        management_fingerprint=management_fingerprint,
        suggested_expectation=expectation,
        base_record_version=0,
        source="provider_intelligence_rule",
        source_rule=OBSERVED_RUNNING_RULE,
        reason=OBSERVED_RUNNING_REASON,
    )
